std/machines/hash/keccakf16_memory.asm

```diff
@@ -0,0 +1,876 @@
+
+machine Keccakf16Memory(mem: Memory, add_sub: AddSub) with
+    latch: first_step,
+    operation_id: operation_id,
+    call_selectors: sel,
+{
+    // ------------- Begin memory read / write ---------------
+    // Additional columns compared to the non-memory version:
+    // - 2 columns (1 high and 1 low) for user input address (of first byte of input).
+    // - 2 columns (1 high and 1 low) for user output address (of first byte of output).
+    // - 98 columns (49 high and 49 low) for computed input/output address of all bytes.
+    // Overall, given that there are 2,600+ columns in the non-memory version, this isn't a huge cost.
+
+    // Methodology description:
+    // 1. User latch in the input and output addresses in the first row of each block.
+    // 2. Input addresses for all bytes are calculated from user input address in the first row using permutation links to add_sub.add operations.
+    // 3. Load all input bytes from memory to the preimage columns.
+    // 4. User output address is kept the same for use in the last row.
+    // 5. Keccak is computed from top to bottom.
+    // 6. Output addresses for all bytes are calculated from user output address in the last row using permutation links to add_sub.add operations.
+    // 7. Store all output bytes from keccak computation columns to memory.
+    // Note that this methodology reuses the same 98 columns of addr_h (49) and addr_l (49) to calculate input and output addresses of all bytes.
+    // However, these 98 columns are only used in the first and last rows of each block.
+    // Essentially, we conduct all memory reads in the first row and all memory writes in the last row.
+    // This might seem wasteful, but it's still cheaper than spreading memory reads/writes over different rows while using as few columns as possible, 
+    // which requires 100 columns to make outputs available in all rows in additional to the memory columns.
+    // This alternative methodology (memory reads/writes over different rows) also doesn't work well with our auto witgen infrastructure, 
+    // because one would need to do memory read -> keccak computation -> memory write as three sequential passes during witgen.
+    // On the contrary, our current methodology performs all memory reads at once in the first row, then immediately does the keccak computation,
+    // and finally performs all memory writes at once in the last row, and thus only requires one pass with auto witgen.
+
+    operation keccakf16_memory<0> input_addr_h, input_addr_l, output_addr_h, output_addr_l, time_step ->;
+
+    // Repeat the time step and input / output address in the whole block.
+    let time_step;
+    unchanged_until(time_step, final_step + is_last);
+
+    // Input address for the first byte of input array from the user.
+    // Used immediately in the first row and therefore doesn't need to be kept constant throughout the block.
+    pol commit input_addr_h;
```

Comment:
```suggestion
    col witness input_addr_h;
```
also below. Or just `let ... ` if that works?

Reply:
I made everything `let` (removed `pol commit` and `col witness`). Actually been pondering why are we supporting all three sytaxes as `let` can just infer the witness type, but regardless I think changing them all to one syntax is cleaner.